module/kops.tf

```diff
@@ -4,11 +4,47 @@ resource "null_resource" "check_kops_version" {
   }
 }
 
+data "template_file" "tmpl_values" {
+  template = "${file("${path.module}/values.yaml.tmpl")}"
+
+  vars {
+    cluster_fqdn          = "${local.cluster_fqdn}"
+    kubernetes_version    = "${var.kubernetes_version}"
+    kops_dns_mode         = "${var.kops_dns_mode}"
+    kops_s3_bucket_id     = "${var.kops_s3_bucket_id}"
+    kubernetes_networking = "${var.kubernetes_networking}"
+    zones                 = "${join(",", data.aws_availability_zones.available.names)}"
+    node_asg_desired      = "${var.node_asg_desired}"
+    master_azs            = "${local.master_azs}"
+    vpc_id                = "${var.vpc_id}"
+    ami_name              = "${local.ami_name}"
+    k8s_apiserver_options = "${join("\n  ", compact(local.k8s_apiserver_options))}"
+  }
+}
+
+resource "null_resource" "generate_values" {
+  triggers {
+    template = "${data.template_file.tmpl_values.rendered}"
+  }
+
+  provisioner "local-exec" {
+    command = "echo \"${data.template_file.tmpl_values.rendered}\" > values.yaml"
+  }
+}
+
+resource "null_resource" "generate_template" {
+  depends_on = ["null_resource.generate_values"]
+
+  provisioner "local-exec" {
+    command = "kops toolbox template --values values.yaml --template ${path.module}/cluster_config.yaml.tmpl --output cluster_config.yaml"
+  }
+}
+
 resource "null_resource" "create_cluster" {
-  depends_on = ["null_resource.check_kops_version"]
+  depends_on = ["null_resource.check_kops_version", "null_resource.generate_template"]
 
   provisioner "local-exec" {
-    command = "kops create cluster --cloud=aws --dns ${var.kops_dns_mode} --authorization RBAC --networking ${var.kubernetes_networking} --zones=${join(",", data.aws_availability_zones.available.names)} --node-count=${var.node_asg_desired} --master-zones=${local.master_azs} --target=terraform --api-loadbalancer-type=public --vpc=${var.vpc_id} --state=s3://${var.kops_s3_bucket_id} --kubernetes-version ${var.kubernetes_version} ${local.cluster_fqdn}"
+    command = "kops create -f cluster_config.yaml --state=s3://${var.kops_s3_bucket_id} && kops create secret --state=s3://vend-test-kops-sydney --name cluster.ap-southeast-2.test.vendhq.com sshpublickey admin -i ~/.ssh/id_rsa.pub && kops update cluster ${var.cluster_fqdn} --state=s3://${var.kops_s3_bucket_id} --target=terraform --yes"
```

Comment:
There is still odd hard coded values in here. e.g. `vend-test-kops-sydney` and `cluster.ap-southeast-2.test.vendhq.com`.

Also, you are still using the `cluster_config.yaml` template by default. This needs to be optional, if enabled.

Reply:
Ops! I must've copy pasted from my test setup and forgotten! Good catch!

Regarding the `cluster_config.yaml` template being optional, as I said above I don't think I can make it optional in a neat fashion as Terraform is limited when it comes to conditionals.
Is there anything in particular that is concerning regarding using the template file?